content/posts/build-a-grafana-dashboard-generator/index.md

```diff
@@ -0,0 +1,401 @@
+a simple process.
+
+=> That’s why we want to use your dashboard as code.
+
+## [Jsonnet](https://github.com/google/go-jsonnet)
+
+A data templating language for app and tool developers
+
+- Generate config data
+- Side-effect free
+- Organize, simplify, unify
+- Manage sprawling config
+
+The example:
+
+- You define a local variable and then that same local variable is referenced later:
+
+```
+local greeting = "hello world!";
+
+{
+ foo: "bar",
+ dict: {
+    nested: greeting
+ },
+} 
+```
+
+=>
+
+```
+{
+ "foo": "bar",
+ "dict": {
+    "nested": "hello world!"
+ }
+} 
+```
+
+- Or, use functions:
+
+```
+// A function that returns an object.
+local Person(name='Alice') = {
+  name: name,
+  welcome: 'Hello ' + name + '!',
+};
+{
+  person1: Person(),
+  person2: Person('Bob'),
+}
+```
+
+=>
+
+```
+{
+  "person1": {
+    "name": "Alice",
+    "welcome": "Hello Alice!"
+  },
+  "person2": {
+    "name": "Bob",
+    "welcome": "Hello Bob!"
+  }
+}
+```
+
+- Patches: You call your Jsonnet function and append to it the snippet of JSON, and it simply adds — overwrites: like
+  so:
+
+```
+dashboard: {
+  new(title, uid): {...}
+}
+
+dashboard.new(...) + {
+ schemaVersion: 22,
+}
+```
+
+=>
+
+```
+{
+  "title": "super dash",
+  "uid": "my-dash",
+  "Panels": [],
+  "schemaVersion": 22
+}
+```
+
+- Imports: Not only can you create functions, but you also can put those functions into files.
+
+```
+{
+  alertlist:: import 'alertlist.libsonnet',
+  dashboard:: import 'dashboard.libsonnet',
+  gauge:: error 'gauge is removed, migrate to gaugePanel',
+  gaugePanel:: import 'gauge_panel.libsonnet',
+}
+```
+
+## [Grafonnet](https://github.com/grafana/grafonnet-lib)
+
+- Is a very simple library that provides you with the basics: creating a dashboard, creating a panel, creating a single
+  stat panel, and so on.
+- See more examples: https://github.com/grafana/grafonnet-lib/tree/master/examples
+- E.g: we have one `dashboard.jsonnet` jsonnet file, maybe call it is main file which will evaluates it to json file
+
+```
+local grafana = import 'grafonnet/grafana.libsonnet';
+local dashboard = grafana.dashboard;
+local annotation = grafana.annotation;
+local template = grafana.template;
+local graphPanel = grafana.graphPanel;
+local prometheus = grafana.prometheus;
+
+dashboard.new(
+  'title',
+  editable=true,
+  refresh='5s',
+  time_from='now-6h',
+  time_to='now',
+  timepicker={},
+  schemaVersion=27,
+  uid='uid',
+)
+.addAnnotation(annotation.default)
+.addTemplate(
+  template.datasource(
+    name='cluster',
+    query='prometheus',
+    current='Thanos',
+    hide='',
+  )
+)
+.addTemplate(
+  template.new(
+    name='namespace',
+    datasource='${cluster}',
+    query={
+        query: 'label_values(grpc_io_server_completed_rpcs, namespace)',
+        refId: 'StandardVariableQuery'
+    },
+    label='Namespace',
+    hide='',
+    refresh='load',
+    definition='label_values(grpc_io_server_completed_rpcs, namespace)'
+  )
+)
+.addPanel(
```

Comment:
Can you add an example image on what will this panel look like in Grafana UI?

Reply:
sure